### PR TITLE
Add API Validators

### PR DESCRIPTION
**Pull request by djimnz**
### Add API Validators

Ensure POST Transactions payloads fulfill the following validations criteria:

- Source and Destinations should match: [a-zA-Z_0-9]+(:[a-zA-Z_0-9]){0,}.
- Amounts should match: positive integer.
- Referenced transaction's assets should match: [A-Z]{1,16}.

### Payload Example - POST Transactions

```
{
    "postings": [
        {
        "source": "world",
        "destination": "central_bank",
        "asset": "GEM",
        "amount": 0 <-------------- VALIDATION WILL FAIL
      },
      {
        "source": "central_bank",
        "destination": "users:001",
        "asset": "GEMw", <-------------- VALIDATION WILL FAIL
        "amount": 1
      }
    ]
}
```

### Response Example - POST Transactions

```
{
    "error": true,
    "error_code": 400,
    "error_message": "Key: 'Transaction.Postings[0].Amount' Error:Field validation for 'Amount' failed on the 'required' tag\nKey: 'Transaction.Postings[1].Asset' Error:Field validation for 'Asset' failed on the 'asset' tag",
    "ok": false
}
```

### CODE REVIEW

File: core/asset.go [hunk 1]
```diff
@@ -1,9 +1,11 @@
...
+	"regexp"
+)
 
-func AssetIsValid(v string) bool {
-	re := regexp.MustCompile("[A-Z]{1,8}")
-
-	return re.Match([]byte(v))
+// AssetIsValid
+func AssetIsValid(asset string) bool {
+	valid, _ := regexp.MatchString("^[A-Z]{1,16}$", asset)
```
Comment (gfyrag) [inline core/asset.go]: Panic if the regexp is not valid seems a good point here.
Even initialize the regexp as global package var would be a good idea. The binary will panic immediately if not valid.
Agree with the fact than today, it works properly, but maybe, the charset could be extended.

File: core/posting.go [hunk 2]
```diff
@@ -1,14 +1,19 @@
...
+
+// Posting struct
 type Posting struct {
-	Source      string `json:"source"`
-	Destination string `json:"destination"`
-	Amount      int64  `json:"amount"`
-	Asset       string `json:"asset"`
+	Source      string `json:"source" validate:"required,source"`
+	Destination string `json:"destination" validate:"required,destination"`
+	Amount      int64  `json:"amount" validate:"required,min=2"`
```
Comment (gfyrag) [inline core/posting.go]: min=2 is by spec?
I think about 1 BTC.

File: core/posting.go [hunk 3]
```diff
@@ -21,3 +26,9 @@ func (ps Postings) Reverse() {
 		ps[opp].Source, ps[opp].Destination = ps[opp].Destination, ps[opp].Source
 	}
 }
+
+// IsValidSourceOrDestination
+func IsValidSourceOrDestination(source string) bool {
+	valid, _ := regexp.MatchString("^[a-zA-Z_0-9]+(:[a-zA-Z_0-9]+){0,}$", source)
```
Comment (gfyrag) [inline core/posting.go]: Same things about regexp.

File: core/validator.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	validator.RegisterValidation("source", validateSourceOrDestination)
+	validator.RegisterValidation("destination", validateSourceOrDestination)
+	validator.RegisterValidation("asset", validateAsset)
+}
+
+// Validate
+func Validate(value interface{}) error {
+	validator := validator.New()
+	register(validator)
+	if err := validator.Var(value, "required,dive"); err != nil {
```
Comment (gfyrag) [inline core/validator.go]: You could register the validator at package level, or pass through the ledger though the resolver to not recreate it each time.

File: core/posting.go (hunk 2 above)
Comment (djimnz) [reply core/posting.go]: Yes sure! I was about to send the commit. I forgot to revert the configuration while testing the validation. 
Thanks

File: core/asset.go (hunk 1 above)
Comment (djimnz) [reply core/asset.go]: The pull request got updated following your recommendation. Feels free to let me know if that way is something you expected to be implemented :)

File: core/validator.go (hunk 4 above)
Comment (djimnz) [reply core/validator.go]: The pull-request got updated following your recommendation. The validator is now injected from the ledger resolver.

**Pull request closed without merging**